File: Xamarin/SampleXamarin.Android/LocationProviderSensorStatus.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

using Microsoft.Azure.SpatialAnchors;

namespace SampleXamarin
{
    public class LocationProviderSensorStatus : SensorStatusModel
    {
        private PlatformLocationProvider locationProvider;

        public LocationProviderSensorStatus(PlatformLocationProvider locationProvider)
        {
            this.locationProvider = locationProvider;
        }

        public SensorStatus GeoLocationStatus
        {
            get
            {
                if (locationProvider == null)
                {
                    return SensorStatus.Indeterminate;
                }
                if (!locationProvider.Sensors.GeoLocationEnabled)
                {
                    return SensorStatus.Blocked;
                }

                var geoLocationStatus = locationProvider.GeoLocationStatus;
                if (geoLocationStatus == GeoLocationStatusResult.Available)
                {
                    return SensorStatus.Available;
                }
                if (geoLocationStatus == GeoLocationStatusResult.DisabledCapability)
                {
                    return SensorStatus.Blocked;
                }
                if (geoLocationStatus == GeoLocationStatusResult.NoGPSData)
                {
                    return SensorStatus.Unavailable;
                }
                return SensorStatus.Indeterminate;
            }
        }

        public SensorStatus WifiSignalStatus
        {
            get
            {
                if (locationProvider == null)
                {
                    return SensorStatus.Indeterminate;
                }
                if (!locationProvider.Sensors.WifiEnabled)
                {
                    return SensorStatus.Blocked;
                }

                var wifiStatus = locationProvider.WifiStatus;
                if (wifiStatus == WifiStatusResult.Available)
                {
                    return SensorStatus.Available;
                }
                if (wifiStatus == WifiStatusResult.DisabledCapability)
                {
                    return SensorStatus.Blocked;
                }
                if (wifiStatus == WifiStatusResult.NoAccessPointsFound)
                {
                    return SensorStatus.Unavailable;
                }
                return SensorStatus.Indeterminate;
            }
        }

        public SensorStatus BluetoothSignalStatus
        {
            get
            {
                if (locationProvider == null)
                {
                    return SensorStatus.Indeterminate;
                }
                if (!locationProvider.Sensors.BluetoothEnabled)
                {
                    return SensorStatus.Blocked;
                }

                var bluetoothStatus = locationProvider.BluetoothStatus;
                if (bluetoothStatus == BluetoothStatusResult.Available)
                {
                    return SensorStatus.Available;
                }
                if (bluetoothStatus == BluetoothStatusResult.DisabledCapability)
                {
                    return SensorStatus.Blocked;
                }
                if (bluetoothStatus == BluetoothStatusResult.NoBeaconsFound)
                {
                    return SensorStatus.Unavailable;
                }
                return SensorStatus.Indeterminate;
            }
        }
    }
}

File: iOS/Swift/SampleSwift/Models/LocationProviderSensorStatus.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

class LocationProviderSensorStatus: SensorStatusModel {
    private weak var locationProvider: ASAPlatformLocationProvider?

    var geoLocationStatus: SensorStatus {
        get {
            if let locationProvider = locationProvider {
                if !locationProvider.sensors.geoLocationEnabled {
                    return .Blocked
                }
                switch (locationProvider.geoLocationStatus){
                case ASAGeoLocationStatusResult.available :
                    return .Available
                case ASAGeoLocationStatusResult.disabledCapability:
                    return .Blocked
                case ASAGeoLocationStatusResult.missingSensorFingerprintProvider:
                    return .Indeterminate
                case ASAGeoLocationStatusResult.noGPSData:
                    return .Unavailable
                }

            }
            return .Indeterminate
        }
    }

    var wifiSignalStatus: SensorStatus {
        get {
            if let locationProvider = locationProvider {
                if !locationProvider.sensors.wifiEnabled {
                    return .Blocked
                }
                switch (locationProvider.wifiStatus){
                case ASAWifiStatusResult.available :
                    return .Available
                case ASAWifiStatusResult.disabledCapability:
                    return .Blocked
                case ASAWifiStatusResult.missingSensorFingerprintProvider:
                    return .Indeterminate
                case ASAWifiStatusResult.noAccessPointsFound:
                    return .Unavailable
                }
            }
            return .Indeterminate
        }
    }

    var bluetoothSignalStatus: SensorStatus {
        get {
            if let locationProvider = locationProvider {
                if !locationProvider.sensors.bluetoothEnabled {
                    return .Blocked
                }
                switch (locationProvider.bluetoothStatus){
                case ASABluetoothStatusResult.available :
                    return .Available
                case ASABluetoothStatusResult.disabledCapability:
                    return .Blocked
                case ASABluetoothStatusResult.missingSensorFingerprintProvider:
                    return .Indeterminate
                case ASABluetoothStatusResult.noBeaconsFound:
                    return .Unavailable
                }
            }
            return .Indeterminate
        }
    }

    init(for locationProvider: ASAPlatformLocationProvider?) {
        self.locationProvider = locationProvider
    }
}
